woocommerce/woocommerce
# Introduce the object cache infrastructure

### All Submissions:


### Changes proposed in this Pull Request:

#### The basics

This pull request introduces the `ObjectCache` class. It's an abstract class that allows to create classes intended to cache any kind of object (this includes arrays and actual PHP objects) under an unique key (which can be an integer or a string) and with a given expiration time (there's a customizable default), as discussed in p7bje6-45j-p2.

At the very least derived classes must implement the `get_object_type` method, which returns a unique identifier (unique across implementations of the class) for the type of object being cached; for example `order` or `product`. There are other protected methods to customize the behavior of the class, but they all have a default implementation.

`ObjectCache` ultimately delegates the actual caching to a class that implemens the (also introduced in this pull request)  `CacheEngine` interface. By default an instance of `WpCacheEngine`, which uses the built-in WordPress caching engine, is used; but that's customizable for each implementation.

The public interface of the class is as one would expect: a `set($id, $object)` method, a `get($id)` method, a `remove($id)` method, a `flush()` method and a `is_cached($id)` method. `get` has actually more (optional) features, more on that soon.

#### Using a custom cache engine

There are two mechanisms for implementations to use a custom cache engine:

* Internally, implementations can override the `get_cache_engine_instance` method.
* Externally, by hooking on the `wc_object_cache_get_engine` filter. This filter receives the default engine to use and the cache object that is triggering the filter.

#### Expirations

All objects are cached with an expiration time, which can range from one second to one month. If no explicit expiration is passed to `set` then the value of `$default_expiration` (one hour by default, overridable by each implementation) will be used.

The actual expiration is expected to be handled by the underlying cache engine, `ObjectCache` will just pass the raw value received by `set`.

#### Validation

Implementations can optionally override the `validate` method to check if the object passed to `set` is valid and return an array of error messages otherwise. If that happens `set` will throw an exception.

#### Serialization and deserialization

Implementations can optionally override the `serialize` and `deserialize` methods:

* `serialize` transforms the object passed to `set` into an array, which is what will be passed to the caching engine. By default it just generates `['data' => $object]`.
* `deserialize` transforms the array returned by the caching engine into the object that will be returned by `get`.  By default it just returns `$serialized['data']`.

Of course the two methods must be symmetric, and implementations should either override both or not override any.

There are also two filters involved in the process:

* `woocommerce_after_serializing_{type}_for_caching`: receives the object and the result of `serialize` and returns what will actually be passed to the caching engine.
* `woocommerce_after_deserializing_{type}_from_cache`: receives the serialized data and the result of `deserialize` and returns what will actually be returned by `get`.

These filters can be used to store some kind of custom metadata together with the actual objects being cached.

#### Object ids inference

`set` can receive null as the id of the object being cached, and by default this will throw an exception. However implementations can override `get_object_id`, which will be called by `set` whenever it receives a null id. If this method is unable to retrieve the id of the passed object it can return null to resort to the default behavior.

#### Caching during retrieval

By default `get` will return null when there isn't any object cached under the passed id. However it's possible to supply code that will retrieve an object from some data store when that happens, this retrieved object will be cached on the spot and then returned by `get`. There are two ways to do that:

* Implementations can override `get_from_datastore`.
* A callback can be passed to `get`.

Both can return null if it's not possible to retrieve an object with the passed id, in order to resort the the default behavior. The callback passed to `get` has precedence over `get_from_datastore`.

`get` accepts an optional expiration parameter to be used whenever a new object is retrieved and cached using this mechanism, of course it works the same way as in `set`. The same applies to validation.

#### Removing objects from cache

The `remove` method will (attempt to) remove a given object from the cache by id, this is done by just passing the request to the cache engine. After that it will trigger the `woocommerce_after_removing_{type}_from_cache` action, which will get the object id as a parameter.

#### Keying and flushing strategy

The underlying cache engine stores objects under a key that is composed by prepending the object id with a prefix. This prefix is retrieved from an option named `wp_object_cache_key_prefix_{type}`; if the option doesn't exist the first time it's needed, the prefix is generated and stored.

The prefix is generated by using the object type and a random string. By default the random string is generated using a combination of `microtime` and `random_bytes`, but implementations can override this via the `get_random_string` method.

What `flush` actually does is to delete the `wp_object_cache_key_prefix_{type}` option, thus effectively rendering the already cached objects inaccessible and forcing the generation of a new prefix the next time it's needed. The inaccessible objects will eventually disappear via the cache engine expiration mechanism.

There's a `woocommerce_after_flushing_{type}_cache` action that is triggered when `flush` is called and gets both the cache object triggering the action and the cache engine as parameters.

### How to test the changes in this Pull Request:

Run the unit tests. Additionally you can create a class that inherits from `ObjectCache`, possibly overriding the protected methods, and see how it behaves.

### Other information:




### FOR PR REVIEWER ONLY:
